File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/preferences/locationathan/athan/AthanVolumeDialog.java
package com.byagowi.persiancalendar.ui.preferences.locationathan.athan;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.view.View;
import android.widget.SeekBar;

import com.byagowi.persiancalendar.utils.Utils;

import java.io.IOException;

import androidx.preference.PreferenceDialogFragmentCompat;

public class AthanVolumeDialog extends PreferenceDialogFragmentCompat {
    private int volume;
    private AudioManager audioManager;
    private Ringtone ringtone;
    private MediaPlayer mediaPlayer;

    @Override
    protected View onCreateDialogView(Context context) {
        final AthanVolumePreference athanPref = (AthanVolumePreference) getPreference();

        audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        if (audioManager != null) {
            audioManager.setStreamVolume(AudioManager.STREAM_ALARM, athanPref.getVolume(), 0);
        }

        Uri customAthanUri = Utils.getCustomAthanUri(context);
        if (customAthanUri != null) {
            ringtone = RingtoneManager.getRingtone(context, customAthanUri);
            ringtone.setStreamType(AudioManager.STREAM_ALARM);
            ringtone.play();
        } else {
            MediaPlayer player = new MediaPlayer();
            try {
                player.setDataSource(context, Utils.getDefaultAthanUri(context));
                player.setAudioStreamType(AudioManager.STREAM_ALARM);
                player.prepare();
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                player.start();
                mediaPlayer = player;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        SeekBar seekBar = new SeekBar(context);
        seekBar.setMax(7);

        volume = athanPref.getVolume();
        seekBar.setProgress(volume);
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                volume = progress;
                audioManager.setStreamVolume(AudioManager.STREAM_ALARM, progress, 0);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                if (ringtone != null && !ringtone.isPlaying()) {
                    ringtone.play();
                }
                if (mediaPlayer != null) {
                    try {
                        if (!mediaPlayer.isPlaying()) {
                            mediaPlayer.prepare();
                            mediaPlayer.start();
                        }
                    } catch (IOException | IllegalStateException e) {
                        e.printStackTrace();
                    }
                }
            }
        });

        return seekBar;
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        final AthanVolumePreference athanPref = (AthanVolumePreference) getPreference();
        if (ringtone != null) {
            ringtone.stop();
        }
        if (mediaPlayer != null) {
            try {
                if (mediaPlayer.isPlaying()) {
                    mediaPlayer.stop();
                    mediaPlayer.release();
                }
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        }
        if (positiveResult) {
            athanPref.setVolume(volume);
        }
    }
}

File: PersianCalendar/src/main/kotlin/com/byagowi/persiancalendar/ui/preferences/locationathan/athan/AthanVolumeDialog.kt
package com.byagowi.persiancalendar.ui.preferences.locationathan.athan

import android.app.Activity
import android.media.AudioManager
import android.media.RingtoneManager
import androidx.core.content.edit
import androidx.core.content.getSystemService
import com.byagowi.persiancalendar.PREF_ATHAN_VOLUME
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.utils.appPrefs
import com.byagowi.persiancalendar.utils.athanVolume
import com.byagowi.persiancalendar.utils.getAthanUri
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.slider.Slider

fun showAthanVolumeDialog(activity: Activity) {
    var volume = activity.athanVolume

    val audioManager = activity.getSystemService<AudioManager>() ?: return
    val originalAlarmVolume = audioManager.getStreamVolume(AudioManager.STREAM_ALARM)
    audioManager.setStreamVolume(AudioManager.STREAM_ALARM, volume, 0)

    val ringtone = RingtoneManager.getRingtone(activity, getAthanUri(activity))
    ringtone?.streamType = AudioManager.STREAM_ALARM
    ringtone?.play()

    val slider = Slider(activity)
    slider.valueFrom = 0f
    slider.valueTo = audioManager.getStreamMaxVolume(AudioManager.STREAM_ALARM).toFloat()
    slider.stepSize = 1f
    slider.value = volume.toFloat()
    slider.addOnChangeListener { _, value, _ ->
        if (ringtone?.isPlaying == false) {
            ringtone.play()
        }
        volume = value.toInt()
        audioManager.setStreamVolume(AudioManager.STREAM_ALARM, volume, 0)
    }

    MaterialAlertDialogBuilder(activity)
        .setTitle(R.string.athan_volume)
        .setView(slider)
        .setPositiveButton(R.string.accept) { _, _ ->
            activity.appPrefs.edit { putInt(PREF_ATHAN_VOLUME, volume) }
        }
        .setNegativeButton(R.string.cancel, null)
        .setOnDismissListener {
            ringtone?.stop()
            audioManager.setStreamVolume(AudioManager.STREAM_ALARM, originalAlarmVolume, 0)
        }
        .show()
}
